test/e2e/expirable-airdrop.spec.ts

```diff
@@ -0,0 +1,188 @@
+import { then, when, given } from '@utils/bdd';
+import { toUnit } from '@utils/bn';
+import { ethers } from 'hardhat';
+import { expect } from 'chai';
+import { BigNumber, Wallet } from 'ethers';
+import { takeSnapshot, time, setBalance, impersonateAccount } from '@nomicfoundation/hardhat-network-helpers';
+import { MerkleTree } from 'merkletreejs';
+
+function getLeaf(address: string, amount: BigNumber) {
+  return Buffer.from(ethers.utils.solidityKeccak256(['address', 'uint256'], [address, amount]).slice(2), 'hex');
+}
+
+function createMerkleRoot(accounts: any[], amount: number) {
+  let toClaim = ethers.utils.parseUnits(String(amount));
+
+  const leaves = accounts.map((x) => {
+    let address = x.address;
+    return getLeaf(address, toClaim);
+  });
+
+  const tree = new MerkleTree(leaves, ethers.utils.keccak256, { sortPairs: true });
+  const root = tree.getHexRoot();
+
+  return [root, tree];
+}
+
+describe('Expirable airdrop', () => {
+  let nowTimestamp: number;
+  const oneMonth: number = 30 * 24 * 60 * 60; // one month in seconds
+  let expirationTimestamp: number;
+  let accounts: any[];
+  let deployer: any;
+  let root: any;
+  let tree: any;
+  let token: any;
+  let expirableAirdrop: any;
+  let airdropAmount: number = 10;
+  let snapshot: any;
+  const randomAddress: string = '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045';
+  const randomAddress2: string = '0x2089035369B33403DdcaBa6258c34e0B3FfbbBd9';
+
+  before(async () => {
+    nowTimestamp = await time.latest();
+    expirationTimestamp = nowTimestamp + 3 * oneMonth; // 3 months
+    [deployer, ...accounts] = await ethers.getSigners();
+
+    [root, tree] = createMerkleRoot(accounts, airdropAmount);
+
+    // deploy erc20 mock
+    const Token = await ethers.getContractFactory('ERC20Mock');
+    token = await Token.deploy('Token', 'TKN');
+
+    // deploy contract
+    const ExpirableAirdrop = await ethers.getContractFactory('ExpirableAirdrop');
+    expirableAirdrop = await ExpirableAirdrop.deploy(deployer.address, token.address, expirationTimestamp, root);
+  });
+
+  when('deposit', () => {
+    let toDeposit: BigNumber;
+
+    given(async () => {
+      toDeposit = toUnit(airdropAmount * (accounts.length + 1));
+
+      // transfer some tokens
+      await token.mint(deployer.address, toDeposit);
+    });
+
+    then('airdrop is deposited', async () => {
+      // approve
+      await token.approve(expirableAirdrop.address, ethers.constants.MaxUint256);
+
+      // deposit
+      await expirableAirdrop.depositTokens(toDeposit);
+
+      // save snapshot for later
+      snapshot = await takeSnapshot();
```

Comment:
It's very hard for me to understand what the tests are doing if we are linking them via the snapshot. If we were to change the order, then they would break, right?

What we normally do is:
- We take a snapshot at the end of the `before`
- We restore it in a `beforeEach`

By doing so, the context is always the same for all tests, they can be re-ordered, and they need to explicitely configure the context that they want to test. It might be a little more verbose, but I personally prefer clarity over shortness in tests